lib/stomp.dart

```diff
@@ -49,60 +41,77 @@ class StompClient {
     }
 
     _handler = StompHandler(
-        config: config.copyWith(onConnect: (_, frame) {
-      if (!_isActive) {
-        config.onDebugMessage(
-            '[STOMP] Client connected while being deactivated. Will disconnect');
-        _handler?.dispose();
-        return;
-      }
-      config.onConnect(this, frame);
-    }, onWebSocketDone: () {
-      config.onWebSocketDone();
-
-      if (_isActive) {
-        _scheduleReconnect();
-      }
-    }));
-    _handler.start();
+      config: config.copyWith(
+        onConnect: (_, frame) {
+          if (!_isActive) {
+            config.onDebugMessage(
+                '[STOMP] Client connected while being deactivated. Will disconnect.');
+            _handler?.dispose();
+            return;
+          }
+          config.onConnect(this, frame);
+        },
+        onWebSocketDone: () {
+          config.onWebSocketDone();
+          if (_isActive) {
+            _scheduleReconnect();
+          }
+        },
+      ),
+    )..start();
   }
 
-  Function({Map<String, String> unsubscribeHeaders}) subscribe(
-      {@required String destination,
-      @required Function(StompFrame) callback,
-      Map<String, String> headers}) {
-    return _handler.subscribe(
-        destination: destination, callback: callback, headers: headers);
+  StompUnsubscribe subscribe({
+    required String destination,
+    required StompFrameCallback callback,
+    Map<String, String>? headers,
+  }) {
+    if (_handler != null) {
```

Comment:
I think we should use `connected` here and also throw an Exception in the else branch. This way the caller knows that a subscription or send message was never really transmitted.

Reply:
Something like this?

```dart
if (connected) {
  return _handler!.subscribe(
    destination: destination,
    callback: callback,
    headers: headers,
  );
}

throw StompClientDisconnectedException();
```

I'm not entirely happy with this, because the `connected` getter hides the null-check from the reader, which makes it not obvious that `_handler!` is correct. And unforunately, because Dart can not promote nullable instance fields automatically to a non-nullable type, we can not get rid of the '!' here.

An alternative could be

```dart
final handler = _handler;
if (handler == null || !handler.connected) {
  throw StompClientDisconnectedException();
}

return handler.subscribe(
  destination: destination,
  callback: callback,
  headers: headers,
);
```

which feels icky to me too. What do you think?

Reply:
Couldn't we maybe just  do `handler!.` since we assume that any other state than non-null should throw an exception either way? This way a null exception is thrown when the handler is null and some sort of WebSocketException (i am not sure which one) would be thrown if the underlying channel/sink is no longer open?

Reply:
Honestly, I'm not sure of the underlying exceptions... I observed in my crashlytics that sometimes `_transmit` fails, because `channel` was null in the `StompHandler`: `NoSuchMethodError: The getter 'sink' was called on null. Receiver: null Tried calling: sink`, but I didn't understand why this would happen exactly (and it didn't really matter to me). This happend if I called `subscribe` in the `onConnected` callback - though it did not reappear after explicitly checking the client for both null and an active connection before calling `subscribe`. I never managed to reproduce this locally.

In the current PR there is a null-check around the `channel` in `_transmit`, which prevents this error (and silently ignores it) - I guess if you remove the null-check it would simply crash with a 'null check operator used on a null value' error instead (in this case). Though, as mentioned, I never could actually make it crash locally, hence I'm pretty much guessing here.

Otherwise, I would still prefer for subscribe/send and ack/nack to either 'ignore the callee', if not connected or otherwise throw an explicit exception that tells me whats wrong (i.e. a StompClientDisconnectedException).

BTW, I just realized that `channel` in `StompHandler` is public, can we make this private?

Reply:
I am just guessing here, but the only way i see channel being null when calling subscribe inside the `onConnected` callback is if the `StompHandler` was replaced in the `StompClient ` after calling `start` on it (maybe calling `activate` twice?). This could maybe be fixed if we dispose the handler before opening a new one (although i am not sure of any ramifcations this might have). But we are getting a bit OT.

I think we should make sure that the user is informed when a call to subscribe/send/ack/nack is made with an inactive/disconnected client and not silently ignore it. For me this means:
- When any of those methods is called before the client was initialized (so before `activate` was called): BadState
- When any of those methods is called and the client is no longer connected: Disconnected

What i am unsure of is if we should supply our own Exceptions or just let dart/WebSocketChannel handle this for us like it was in the code before (NoSuchMethodError and whatever Exception WebSocketChannel throws when we try add something to the sink while it is closed. I would assume a BadState)

Reply:
Your arguments make sense. I guess we probably then need two exceptions:
1. BadStateException: Calling send/subscribe/ack/nack while `handler` is null. The only reason for the handler being null at that point is that activate wasn't called or if the beforeConnect future is still running.
2. StompClientDisconnectedException (or something along those lines): Calling send/subscribe/ack/nack while being disconnected. This should probably be thrown inside the `_transmit` method.

Reply:
Perfect. Thanks again for your effort!

Reply:
So, I justed looked into it, and for now I went with a single `StompBadStateException`:

* In `_transmit` I can not check if the client is connected because `_transmit` itself is used to establish the connection between the client and the server. Of course, I can check whether the websocket channel is null, but this is really, as you mentioned before, a 'bad state' that should not actually happen.
* I got rid of the idea of a separate `DisconnectedException` after trying it out, because all exceptions are ultimately about the connection state, hence having an extra exception type does not seem semantically meaningful. Theres nothing different to act on. The only thing it does is force the callee to catch two separate exceptions instead of one.
* Basically, I wrapped `channel!.sink.add` in a try-catch and convert all exceptions into a `StompBadStateException`. In the `StompClient` I throw the exception for when the `_handler` is null.
* I also tried to add meaningful error messages and added two test cases for the handler and the client.

Please have a look and tell me what you think. :)

PS: I also made `channel` private in the `StompHandler`.

Reply:
Very nice. Thank you! I will merge this PR and release a new version this weekend or early next week.

Reply:
Thank you too! 👍🏻